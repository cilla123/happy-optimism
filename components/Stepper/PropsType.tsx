export default interface PropsType {
  theme?: 'default' | 'primary' | 'success' | 'warning' | 'error';
  shape?: 'radius' | 'circle';
  value?: number;
  defaultValue?: number;
  step: number;
  min: number;
  max: number;
  disabled?: boolean;
  onInputChange?: (value: number) => void;
  onChange?: (value: number) => void;
}
