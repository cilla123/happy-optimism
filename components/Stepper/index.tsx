import React, { PureComponent } from 'react';
import classnames from 'classnames';
import PropsType from './PropsType';
import Icon from '../Icon';

const getValue = (props, defaultValue) => {
  if ('value' in props) {
    return props.value;
  }
  if ('defaultValue' in props) {
    return props.defaultValue;
  }
  return defaultValue;
};

export interface StepperProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class Stepper extends PureComponent<StepperProps, any> {

  static defaultProps = {
    prefixCls: 'za-stepper',
    theme: 'primary',
    disabled: false,
    step: 1,
  };

  constructor(props) {
    super(props);
    this.state = {
      value: getValue(props, 0),
      lastValue: getValue(props, 0),
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('value' in nextProps) {
      this.setState({
        value: Number(getValue(nextProps, 0)),
        lastValue: Number(getValue(nextProps, 0)),
      });
    }
  }

  onInputChange = (value) => {
    value = Number(value);
    const { onInputChange } = this.props;
    this.setState({ value });
    if (typeof onInputChange === 'function') {
      onInputChange(value);
    }
  }

  onInputBlur = (value) => {
    const { min, max, onChange } = this.props;
    value = Number(value);
    if (value === '' || isNaN(value)) {
      value = this.state.lastValue;
    }
    if (min !== null && value < min) {
      value = min;
    }
    if (max !== null && value > max) {
      value = max;
    }
    this.setState({
      value,
      lastValue: value,
    });
    if (typeof onChange === 'function') {
      onChange(value);
    }
  }

  onSubClick = () => {
    const { step } = this.props;
    const { value } = this.state;
    if (this.isSubDisabled()) {
      return;
    }

    const newValue = Number(value) - step;
    this.onInputBlur(newValue);
  }

  onPlusClick = () => {
    const { step } = this.props;
    const { value } = this.state;
    if (this.isPlusDisabled()) {
      return;
    }

    const newValue = Number(value) + step;
    this.onInputBlur(newValue);
  }

  isSubDisabled = () => {
    const { min, disabled } = this.props;
    const { value } = this.state;

    if (min === null) {
      return false;
    }
    return (value <= min) || disabled;
  }

  isPlusDisabled = () => {
    const { max, disabled } = this.props;
    const { value } = this.state;

    if (max === null) {
      return false;
    }
    return (value >= max) || disabled;
  }

  render() {
    const { prefixCls, className, theme, shape, disabled } = this.props;
    const { value } = this.state;

    const cls = classnames(`${prefixCls}`, className, {
      [`theme-${theme}`]: !!theme,
      [`shape-${shape}`]: !!shape,
      disabled,
    });

    const subCls = classnames(`${prefixCls}-sub`, {
      disabled: this.isSubDisabled(),
    });

    const plusCls = classnames(`${prefixCls}-plus`, {
      disabled: this.isPlusDisabled(),
    });

    return (
      <span className={cls}>
        <span className={subCls} onClick={this.onSubClick}><Icon type="minus" /></span>
        <input
          className={`${prefixCls}-input`}
          type="tel"
          value={value}
          onChange={e => this.onInputChange(e.target.value)}
          onBlur={() => this.onInputBlur(value)}
        />
        <span className={plusCls} onClick={this.onPlusClick}><Icon type="add" /></span>
      </span>
    );
  }
}
