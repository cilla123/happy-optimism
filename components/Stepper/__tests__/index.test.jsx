import React from 'react';
import { render, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Stepper from '../index';

describe('Stepper', () => {
  it('renders correctly', () => {
    const onChange = jest.fn();
    const wrapper = render(
      <Stepper onChange={onChange} />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('defaultValue', () => {
    const wrapper = shallow(
      <Stepper defaultValue={2} />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('receive new value', () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Stepper onChange={onChange} />
    );
    wrapper.setProps({ value: 10 });
  });

  it('onChange and onInputChange', () => {
    const onChange = jest.fn();
    const onInputChange = jest.fn();
    const wrapper = shallow(
      <Stepper onChange={onChange} onInputChange={onInputChange} />
    );
    wrapper.find('input').simulate('change', { target: { value: 10 } });
    expect(onInputChange).toBeCalled();
    wrapper.find('input').simulate('blur');
    expect(onChange).toBeCalledWith(10);
    wrapper.find('.za-stepper-sub').simulate('click');
    expect(onChange).toBeCalledWith(9);
    wrapper.find('.za-stepper-plus').simulate('click');
    expect(onChange).toBeCalledWith(10);
  });
});
