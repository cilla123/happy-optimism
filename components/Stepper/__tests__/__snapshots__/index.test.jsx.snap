// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stepper defaultValue 1`] = `
<span
  className="za-stepper theme-primary"
>
  <span
    className="za-stepper-sub"
    onClick={[Function]}
  >
    <Icon
      prefixCls="za-icon"
      type="minus"
    />
  </span>
  <input
    className="za-stepper-input"
    onBlur={[Function]}
    onChange={[Function]}
    type="tel"
    value={2}
  />
  <span
    className="za-stepper-plus"
    onClick={[Function]}
  >
    <Icon
      prefixCls="za-icon"
      type="add"
    />
  </span>
</span>
`;

exports[`Stepper renders correctly 1`] = `
<span
  class="za-stepper theme-primary"
>
  <span
    class="za-stepper-sub"
  >
    <i
      class="za-icon za-icon-minus"
    />
  </span>
  <input
    class="za-stepper-input"
    type="tel"
    value="0"
  />
  <span
    class="za-stepper-plus"
  >
    <i
      class="za-icon za-icon-add"
    />
  </span>
</span>
`;
